cgi-bin/DW/Controller/Customize.pm

```diff
@@ -0,0 +1,654 @@
+            $queryargs->{show} = LJ::eurl( $post->{show} );
+            $queryargs->{page} = 1;
+        }
+        elsif ( $post->{search} ) {
+            $queryargs->{search} = LJ::eurl( $post->{search} );
+        }
+        elsif ( $post->{which_title} ) {
+            my $eff_val = LJ::text_trim( $post->{title_value}, 0, LJ::std_max_length() );
+            $eff_val = "" unless $eff_val;
+            $u->set_prop( $post->{which_title}, $eff_val );
+        }
+        elsif ( $post->{apply_layout} ) {
+
+            set_layout(
+                {
+                    layout_choice     => $post->{layout_choice},
+                    layout_prop       => $post->{layout_prop},
+                    show_sidebar_prop => $post->{show_sidebar_prop},
+                    u                 => $u
+                }
+            );
+        }
+        my $redir = LJ::create_url( $url, args => $queryargs, no_blank => 1 );
+        return $r->redirect($redir);
+
+    }
+
+    $vars->{theme_data}   = get_themechooser_data($queryargs);
+    $vars->{layout_data}  = get_layout_data();
+    $vars->{current_data} = get_current_data(0);
+
+    # Now we tell it what template to render and pass in our variables
+    return DW::Template->render_template( 'customize/index.tt', $vars );
+
+}
+
+## Handlers for RPC endpoints
+
+sub themechooser_handler {
+    my ( $ok, $rv ) = controller( authas => 1 );
+    return $rv unless $ok;
+
+    # gets the request and args
+
+    my $r        = DW::Request->get;
+    my $args     = $r->post_args;
+    my $themeid  = $args->{apply_themeid};
+    my $layoutid = $args->{apply_layoutid};
+    my $getargs  = make_queryargs($args);
+
+    # apply the new theme selected
+
+    set_theme( apply_themeid => $themeid, apply_layoutid => $layoutid );
+
+    my $theme_data = get_themechooser_data($getargs);
+    my $theme_html = DW::Template->template_string( 'customize/themechooser.tt',
+        { theme_data => $theme_data, qargs => $getargs } );
+    my $layout_data  = get_layout_data();
+    my $current_data = get_current_data(0);
+
+    $r->print(
+        to_json(
+            {
+                theme_html   => $theme_html,
+                layout_data  => $layout_data,
+                current_data => $current_data
+            }
+        )
+    );
+    return $r->OK;
+
+}
+
+sub filter_handler {
+
+    # gets the request and args
+    my ( $ok, $rv ) = controller( authas => 1, form_auth => 1 );
+    return $rv unless $ok;
+
+    my $r         = $rv->{r};
+    my $args      = $r->get_args;
+    my $queryargs = make_queryargs($args);
+
+    my $theme_data = get_themechooser_data($queryargs);
+    my $theme_html = DW::Template->template_string( 'customize/themechooser.tt',
+        { 'theme_data' => $theme_data, qargs => $queryargs } );
+    my $current_data = get_current_data(0);
+
+    $r->print( to_json( { theme_html => $theme_html, current_data => $current_data } ) );
+    return $r->OK;
+}
+
+sub layoutchooser_handler {
```

Comment:
Don't you need 

```perl
    my ( $ok, $rv ) = controller( authas => 1, form_auth => 1 );
    return $rv unless $ok;
```

to be safe?

Reply:
So because this and the journaltitles_handler are only hit via JS, and don't rewrite large swathes of DOM, the form_auth value goes stale after the first request using it. I feel like these two are fairly low risk for CSRF, though?